modules/core/src/test/java/org/apache/ignite/internal/processors/cache/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,318 @@
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }, NODES_CNT, "worker").get(30_000L);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverUnfair() throws Exception {
+        testReentrantLockMultinodeFailover(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverFair() throws Exception {
+        testReentrantLockMultinodeFailover(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailover(final boolean fair) throws Exception {
+        List<IgniteInternalFuture<Void>> futs = new ArrayList<>(NODES_CNT);
+
+        for (int i = 0; i < NODES_CNT; i++) {
+            final int inx = i;
+            final boolean flag = RND.nextBoolean();
+
+            futs.add(GridTestUtils.runAsync(new Callable<Void>() {
+                @Override public Void call() throws Exception {
+                    IgniteLock lock = createReentrantLock(inx, fair ? "lock1" : "lock2", fair);
+
+                    UUID id = grid(inx).cluster().localNode().id();
+
+                    lock.lock();
+
+                    if (flag) {
+                        try {
+                            grid(inx).close();
+                        }
+                        catch (Exception ignored) {
+                            lock.unlock();
+                        }
+
+                        return null;
+                    }
+
+                    try {
+                        assertTrue(lock.isLocked());
+
+                        Thread.sleep(1_000L);
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+
+                    return null;
+                }
+            }));
+        }
+
+        for (IgniteInternalFuture<?> fut : futs)
+            fut.get(60_000L);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocksUnfair() throws Exception {
+        testReentrantLockMultinodeFailoverMultilocks(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocksFair() throws Exception {
+        testReentrantLockMultinodeFailoverMultilocks(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocks(final boolean fair) throws Exception {
+        List<IgniteInternalFuture<Void>> futs = new ArrayList<>(NODES_CNT * 2);
+
+        for (int i = 0; i < NODES_CNT; i++) {
```

Comment:
GridTestUtils.runMultiThreaded()

Reply:
In Java 7 hard to send threadsafe boolean random number in GridTestUtils.runMultiThreaded().

Reply:
use ThreadLocalRandom.current().nextBoolean() inside every thread

Reply:
ThreadLocalRandom from Java 8.

Reply:
fixed